mem-store.ts

```diff
@@ -109,96 +307,40 @@ function mem_store(this: any, options: any) {
           mement = Object.assign(prev || {}, mement)
         }
 
-        return upsertIfRequested(msg, function(err: any, ctx: UpsertIfRequestedContext): any {
-          if (err) {
-            return reply(err)
-          }
-
-          const { did_update } = (ctx as any)
-
-          if (did_update) {
-            const { out } = (ctx as any)
-            return reply(null, out)
-          }
-
-
-          prev = entmap[base][name][mement.id] = mement
-
-          seneca.log.debug(function() {
-            return [
-              'save/' + (isNewEntityBeingCreated(msg) ? 'insert' : 'update'),
-              ent.canon$({ string: 1 }),
-              mement,
-              desc,
-            ]
-          })
-
-          return reply(null, ent.make$(prev))
-        })
-
-
-        type UpsertIfRequestedContext = { did_update: boolean, out?: any } | undefined
-
-        type UpsertIfRequestedCallback = (
-          err: Error | null,
-          result?: UpsertIfRequestedContext
-        ) => any
-
-        function upsertIfRequested(msg: any, reply: UpsertIfRequestedCallback) {
-          // This is the query passed to the .save$ method.
-          //
-          const query_for_save = msg.q
-
-
-          if (isNewEntityBeingCreated(msg) && Array.isArray(query_for_save.upsert$)) {
-            const upsert_on = Common.clean(query_for_save.upsert$)
-
-
-            if (upsert_on.length > 0) {
-              if (!(base in entmap)) {
-                return reply(null, { did_update: false, out: null })
-              }
-
-              if (!(name in entmap[base])) {
-                return reply(null, { did_update: false, out: null })
-              }
-
+        if (intern.is_upsert_requested(msg)) {
+          const upsert_on = seneca.util.clean(msg.q.upsert$)
 
-              const collection = entmap[base][name]
-              const docs = Object.values(collection)
-              const public_entdata = msg.ent.data$(false)
+          if (upsert_on.length > 0) {
+            const public_entdata = ent.data$(false)
```

Comment:
why is a copy needed here?

Reply:
which private fields?
we should really try hard to avoid another copy here

Reply:
> which private fields?
For example given the following code:
```javascript
seneca.make('race_car_driver')
  .data$({ first_name: 'Richard', psst$: 'psst' })
  .save$({ upsert$: ['psst$'] }, done)
```
-we cannot match on `psst$`. As such, in the code snippet above, no match can ever happen.

The following code features the same scenario, except `psst` is no longer a private field;
```javascript
seneca.make('race_car_driver')
  .data$({ first_name: 'Richard', psst: 'psst' })
  .save$({ upsert$: ['psst'] }, done)
```
-we can match on `psst`. As such, in the code snippet above, a match may happen.

Reply:
> we should really try hard to avoid another copy here

Alright, I'll see what can be done.